server/models/questions/createQuestion.js

```diff
@@ -0,0 +1,122 @@
+ * @param {String} param0.name
+ * @param {String} param0.problem_statement
+ * @param {String} param0.type
+ * @param {String} param0.input_format
+ * @param {String} param0.output_format
+ * @param {String} param0.constraints
+ * @param {Array} param0.options
+ * @param {Number} param0.correct
+ * @param {String} param0.difficulty
+ * @param {Array} param0.tags
+ * @return {Promise}
+ */
+
+function createQuestion({
+  username,
+  name,
+  problem_statement: problemStatement,
+  type,
+  input_format: inputFormat,
+  output_format: outputFormat,
+  constraints,
+  options,
+  correct,
+  difficulty,
+  tags,
+}) {
+  return new Promise((resolve, reject) => {
+    pool.getConnection((err, connection) => {
+      if (err) {
+        return reject(err)
+      }
+      connection.beginTransaction((err) => {
+        if (err) {
+          connection.release()
+          return reject(err)
+        }
+        connection.query(
+          `INSERT INTO questions(creator,type,name,problem_statement,input_format,
+            output_format,constraints,options,correct,difficulty)
+            VALUES(?,?,?,?,?,?,?,?,?,?)`,
+          [
+            username,
+            type,
+            name,
+            problemStatement,
+            inputFormat,
+            outputFormat,
+            constraints,
+            JSON.stringify(options),
+            correct,
+            difficulty,
+          ],
+          (error, results) => {
+            if (error || !results) {
+              connection.release()
+              return reject(error)
+            }
+
+            const questionId = results.insertId
+
+            if (!tags || !tags.length) {
+              return connection.commit((error) => {
+                if (error) {
+                  return connection.rollback(() => {
+                    connection.release()
+                    return reject(error)
+                  })
+                }
+                connection.release()
+                return resolve({
+                  message: 'Question created successfully',
+                  questionId,
+                })
+              })
+            }
+
+            let query = `INSERT INTO questions_tags(question_id, tag_id) VALUES`
+            const queryArr = []
+
+            tags.forEach((tag, idx) => {
+              query += `(?,?)`
+              if (idx !== tags.length - 1) {
+                query += `,`
+              }
+              queryArr.push(questionId, tag)
+            })
+
+            connection.query(query, queryArr, (error) => {
+              if (error) {
+                return connection.rollback(() => {
+                  connection.release()
+                  return reject(error)
+                })
+              }
+              return connection.commit((error) => {
+                if (error) {
+                  return connection.rollback(() => {
+                    connection.release()
+                    return reject(error)
```

Comment:
We need to also update the `questions_editors_map` table also in the same transaction, I suppose. i.e. Insert the creator also in this table.

Reply:
I thought of doing it..But later changed my mind..because a creator can edit questions..right..so by default it is known that he is a creator..

Reply:
This seems to be data redundancy. But it will help us in the remaining endpoints
For remaining endpoints, like update a question, etc. we would need to verify whether the user is a editor or not. If we query 2 tables `questions` and `questions_editors` both, it would work much slower. 
Actually, there's a trade off here. And in our case, the create operation would be much less frequent then update or other requests. So, it is better to insert the creator in `questions_editors_map`

Reply:
same thing is done in `user_groups` i suppose

Reply:
there is a column `is_group_moderator` in that table only

Reply:
while updating a question, we would first look at creator and see if it matches if not only then questions_editors will be queried..so I guess it is not required

Reply:
We may later test the execution times

Reply:
For now, this also seems fine.

Reply:
This has been implemented